testonly/mdm/mdm.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mdm provides test-only code for checking the merge delay of a
+// Trillian log.
+package mdm
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian"
+	"github.com/google/trillian/client"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/types"
+)
+
+// MergeDelayMonitor submits leaves to a Log and measures merge delay.
+type MergeDelayMonitor struct {
+	client []*client.LogClient
+	logID  int64
+	opts   MergeDelayOptions
+}
+
+// MergeDelayOptions holds the parameters for a MergeDelayMonitor.
+type MergeDelayOptions struct {
+	ParallelAdds  int
+	LeafSize      int
+	NewLeafChance int // percentage
+	EmitInterval  time.Duration
+	MetricFactory monitoring.MetricFactory
+}
+
+// NewMonitor creates a MergeDelayMonitor instance for the given log ID, accessed
+// via the cl client.
+func NewMonitor(ctx context.Context, logID int64, cl trillian.TrillianLogClient, acl trillian.TrillianAdminClient, opts MergeDelayOptions) (*MergeDelayMonitor, error) {
+	if opts.MetricFactory == nil {
+		opts.MetricFactory = monitoring.InertMetricFactory{}
+	}
+	if opts.EmitInterval == 0 {
+		opts.EmitInterval = 10 * time.Second
+	}
+	metricsOnce.Do(func() { initMetrics(opts.MetricFactory) })
+
+	tree, err := acl.GetTree(ctx, &trillian.GetTreeRequest{TreeId: logID})
```

Comment:
'acl' is potentially confusing but I don't think it's too bad here.

Reply:
Renamed to `adminCl` here and in mdmtest/main.go.